File: frontend/src/services/tkpi.js
import { apiGet } from './api';

/**
 * Search TKPI database for food items
 * @param {string} query - Search query
 * @param {number} limit - Maximum results (default: 10)
 * @param {boolean|null} fuzzy - Override fuzzy mode:
 *   true  → force token-based fuzzy search
 *   false → force exact-phrase match
 *   null/undefined → use server-side TKPI_FUZZY_SEARCH feature flag
 * @returns {Promise<Array>} Search results [{id, name}]
 */
export async function searchTkpi(query, limit = 10, fuzzy = null) {
  if (!query || query.trim().length === 0) {
    return [];
  }

  let url = `/api/v1/tkpi/search?q=${encodeURIComponent(query)}&limit=${limit}`;
  if (fuzzy !== null && fuzzy !== undefined) {
    url += `&fuzzy=${fuzzy ? 'true' : 'false'}`;
  }

  return apiGet(url);
}

/**
 * Get detailed nutrition info from TKPI by ID
 * @param {number} foodId - TKPI food ID
 * @returns {Promise<Object>} Detailed nutrition information
 * Response format:
 * {
 *   id: number,
 *   name: string,
 *   nutrition: {energi_kal, protein_g, lemak_g, karbo_g, serat_g}
 * }
 */
export async function getTkpiDetail(foodId) {
  return apiGet(`/api/v1/tkpi/${foodId}`);
}

/**
 * Fetch list of foods the YOLO model can currently detect.
 * Returns { total, foods: [{ yolo_label, tkpi_food_id, name, tkpi_code }] }
 */
export async function fetchDetectableFoods() {
  return apiGet('/api/v1/detectable-foods');
}

/**
 * Fetch detectable foods and return a Set of TKPI food IDs for fast lookup.
 * @returns {Promise<Set<number>>}
 */
export async function getDetectableFoodIds() {
  const data = await fetchDetectableFoods();
  return new Set((data.foods || []).map(f => f.tkpi_food_id));
}

File: backend/app/api/v1/endpoints/tkpi.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from app.db.session import get_db
from app.services.tkpi_service import search_tkpi_foods, get_tkpi_food_by_id
from app.schemas.tkpi import TKPISearchItem, TKPIDetail
from app.schemas.common import NutritionInfo
from app.core.config import settings


router = APIRouter()


@router.get("/search", response_model=List[TKPISearchItem])
async def search_tkpi(
    q: str = Query(..., description="Search query"),
    limit: int = Query(10, ge=1, le=50, description="Maximum results"),
    fuzzy: Optional[bool] = Query(
        None,
        description=(
            "Enable token-based fuzzy search. "
            "Overrides the TKPI_FUZZY_SEARCH feature flag when provided. "
            "Pass ?fuzzy=false to force exact-phrase mode on a specific request."
        ),
    ),
    db: Session = Depends(get_db)
):
    """
    Search TKPI foods by name.
    Returns minimal information (id, name) for autocomplete/search results.

    Fuzzy mode (token-based) can be controlled globally via the
    TKPI_FUZZY_SEARCH environment variable, or per-request via ?fuzzy=true/false.
    """
    # Per-request param takes priority; fall back to global feature flag
    use_fuzzy = fuzzy if fuzzy is not None else settings.TKPI_FUZZY_SEARCH
    foods = search_tkpi_foods(db, q, limit, fuzzy=use_fuzzy)
    return foods


@router.get("/{food_id}", response_model=TKPIDetail)
async def get_tkpi_detail(
    food_id: int,
    db: Session = Depends(get_db)
):
    """
    Get detailed TKPI food information including full nutrition data.
    """
    food = get_tkpi_food_by_id(db, food_id)
    
    if not food:
        raise HTTPException(status_code=404, detail=f"TKPI food {food_id} not found")
    
    return TKPIDetail(
        id=food.id,
        name=food.name,
        nutrition=NutritionInfo(
            energi_kal=food.energi_kal,
            protein_g=food.protein_g,
            lemak_g=food.lemak_g,
            karbo_g=food.karbo_g,
            serat_g=food.serat_g
        )
    )
